### ISSUE TITLE
Black borders on mutate (image resize with background color)

### ISSUE THREAD

**Issue by romanov** [SixLabors/ImageSharp#1986]
### Prerequisites

- [+] I have written a descriptive issue title
- [+] I have verified that I am running the latest version of ImageSharp
- [+] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [+] I have searched [open](https://github.com/SixLabors/ImageSharp/issues) and [closed](https://github.com/SixLabors/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
Hello,

i'm trying to resize the following test image (1500x1500) to 250x350px and keep white background:

https://ibb.co/Yk6kqBc

### Steps to Reproduce

1. Upload the image above to the local folder
2. Encode it with the following code

```
var encoder = new WebpEncoder
{
    Quality = 100,
    Method = WebpEncodingMethod.BestQuality,
};

var resizeOptions = new ResizeOptions
{
    Mode = ResizeMode.Pad,
    Size = new Size(250, 350)
};

using var image = Image.Load(fileName, out var imageFormat);
image.Mutate(x => x.Resize(resizeOptions).BackgroundColor(Color.White));
image.SaveAsWebp(newFileName, encoder);
```

Expected result (green box in the middle, white background):

![expected](https://user-images.githubusercontent.com/644179/153256802-797341cb-9c75-448b-ac60-567bb37ea669.jpg)

Actual result (green box in the middle, black background):

![actual](https://user-images.githubusercontent.com/644179/153256887-52586ed4-8a25-446b-8062-eaae8411d5d4.jpg)

### System Configuration
- ImageSharp version: 2.0.0
- Other ImageSharp packages and versions: n/a
- Environment (Operating system, version and so on): Microsoft Windows [Version 10.0.19043.1466]
- .NET Framework version: NET 6 Console app & ASP.NET CORE
- Additional information: n/a

**Comment by brianpopow** [comment]
I cant reproduce it, looks as expected with white border.

P.S. in your above example you are not actually using the encoder. You have to pass the Encoder to `Save` as a parameter:

```
image.Save("output.webp", encoder);
```

**Comment by romanov** [comment]
Yes, i provide encoder in Save (fixed code above).
I made a test solution, can u test it?

[ImageMutateTest.zip](https://github.com/SixLabors/ImageSharp/files/8036381/ImageMutateTest.zip)

**Comment by brianpopow** [comment]
I was able to reproduce it, but I see the issue also for png.

Its seems to be an issue with applying the background, not the encoding of the image.

It seems to work with `Image.Load<Rgba32>`, but loading the image with `Image.Load` (without specifying a pixel type), does not.

**Comment by JimBobSquarePants** [comment]
That's expected. `Pad` uses `BackgroundColor` which requires an alpha component. 

https://github.com/SixLabors/ImageSharp/blob/a844c78d31eaef20467b1ae8adb4beb2988cf006/src/ImageSharp/Processing/Extensions/Transforms/PadExtensions.cs#L32-L40

If you want to set that color you need to use the generic API and decode to an pixel format containing an alpha component.

**Comment by romanov** [comment]
> If you want to set that color you need to use the generic API and decode to an pixel format containing an alpha component.
Maybe, you should add some comment or warning for mutation methods (forcing people to use generic)? Because, i've tried yesterday to convert images and spent like 20 minutes. Never knew that i need to use Image.Load<Rgba32>(...) instead of Image.Load(...)

**Comment by JimBobSquarePants** [comment]
Maybe you could?

**Comment by tocsoft** [comment]
@JimBobSquarePants @antonfirsov ☝️ this is why i warned against us decoding (when using non-generic) Image into a pixel format that doesn't support the full Rgba32 color space (or at least having an alpha channel) 

This 'unexpected behaviour' will likely show up again and again... it doesn't help this is also a breaking behaviour change between v1 and v2.

This change we made now means that different APIs behave different just based on the source image loaded and not on the code someone defined created. One of the major drivers of having the non-generic image was to simplify users interactions with images, we have made it more difficult the only way you can have consistent processing is decode into `Image<Rgba32>` which defeats most of the point of having the non-generic apis in the first place.

**Comment by antonfirsov** [comment]
> Why not introduce `forceRgba32` set to `true` in the `Configuration` then?

I had exactly the same idea, but got pushback 😄 Maybe we can re-evaluate. It could be a very simple opt-out from the optimization without forcing the user to go generic.

@tocsoft in general, I still believe this is something we can explain in conceptual docs rather than penalize every single user's memory footprint with an alpha channel.

**Comment by br3aker** [comment]
> I still believe this issue is something we can address by explaining it in conceptual docs rather than penalize every single user's memory footprint. Most users wont face similar problems.

If it's only a resize type problem - we can solve it way simplier: decode-resize-encode API.

Jpeg will hopefully get that fancy downscaling decoding from me and we would need an API for any image format anyway so user won't need to call `Identify` and then `if-else` for jpeg specific path, maybe we can fit this specific case in that API?

**Comment by JimBobSquarePants** [comment]
We actually have all the information at hand to perform a version of pad with background color that works for pixel formats without an alpha component. ResizeHelper gives us the size and rectangle, we can simply fill around when required. This is the only method I can think of that is an issue. (BackgroundColor is not, since the requirement for an alpha component should be understood)

I don’t see the non-generic api behaviour as an issue that can’t be fixed by documentation. At least it doesn’t throw like some System.Drawing methods.

**Comment by JimBobSquarePants** [comment]
I think target size decoding should be specific to JpegDecoder btw. I’d rather keep that kind of functionality low level and not try to abstract it across image formats

**Comment by tocsoft** [comment]
The issue isn't just resize its also Drawing as a whole, anyone wanting to draw with transparency (i.e. crating rounded core avatar images etc, just alpha pixel blending in general will also be effected. 

My personal recommendation would be path of least surprise (as people will not read the docs to learn whey it not working they will just give up as its buggy and not even tell us), I would switch the default decode back to `Rgba32` but include a Configuration flag for those people that aren't impacted by the change and want the memory reduction can enable the optimisation. 

It not like this change even improved anything for `ImageSharp.Web` (the exact sort of place that would benefit the most from it) because we explicitly use `Image<Rgba32>` over there

**Comment by br3aker** [comment]
> I think target size decoding should be specific to JpegDecoder btw. I’d rather keep that kind of functionality low level and not try to abstract it across image formats

I'll open specific discussion a bit later cause I don't understand the reason it being so.

> I think anyone expecting to work with alpha blending should be expecting to work with the generic API and with a pixel format containing an alpha component. Our previously discussed issues around Color.ToPixel and Rgb24 highlighted the importance of that.

I wasn't involved in previous discussions so sorry if this point was disscussed earlier.

Using explicit `Image.Load<Rgba32>(...)` just to enable alpha channel working is counter-intuitive, in my opinion. If code doesn't work with any pixels then it most likely won't have an explicit pixel type.

**Comment by JimBobSquarePants** [comment]
@br3aker Rethinking the first point it would be possible using default interface implementations but I don't think that can be a V2 thing since any interface change would still be considered breaking.

Regarding the alpha component comment. The issue can be reduced to this. 

``` c#
Color c = Color.FromRgba(255, 255, 255, 128);
Rgb24 r = c.ToPixel<Rgb24>();
r.Dump(); // 255, 255, 255.
```

All our brushes for drawing operations take a `Color` parameter and, should they generate a canvas using a pixel format containing no alpha component, certain operations containing alpha blending (porter duff) lead to results that less experienced developers do not expect.

**Comment by tocsoft** [comment]
It wont just be alpha channel really its about full fidelity color/alpha... If a user decodes an image uploaded/provided by a third party but they want to draw a fancy frame.. maybe has transparent bits maybe has different colored overlays etc ... that should all just work with the 'simple' API.

We can't forget a high majority of our users (on the drawing side at least) will not, and have no need to, care about pixel formats. All they care about is; I can load any image, I can manipulate it simply, and then when I save it it looks like I expect. That should be our primary goal for usage, anything that breaks that experience (no matter the perceived benefits, be that memory or speed or anything else) should be opt it.

**Comment by JimBobSquarePants** [comment]
@antonfirsov Ok so based on your interpretation of the below.

>If code doesn't work with any pixels then it most likely won't have an explicit pixel type.

If someone currently wants to set the background color of an image following a resize by using separate operations they must provide an image with a pixel format containing an alpha component. `BackgroundColor` requires that component and I would expect users to make the least amount of effort and attempt to understand why that is required. The fact that 24 bit per component Pngs for example would fail deliver the "expected" result despite users associating the Png format with transparency should not surprise a user. We can document that the method requires an alpha component or at least explain in the docs how it works.

Back to the problem at hand.

`Pad` is the only process we have built in the main library that fails outright (_delivers a result so unexpected it could be rightly interpreted as a bug_) when given an image decoded to a pixel format without an alpha component. It combines `Resize` with `BackgroundColor` under the guise of a single operation. We can fix that problem without breaking changes by adding a `FillColor` property to `ResizeOptions` and filling bounds within the padded sections of the image.

For drawing operations when someone is not directly touching the pixel buffer and is relying on our high level APIs, for example, Fill with alpha, then I would say they must provide a pixel format with an alpha component or they'll get unexpected (to them but absolutely expected to us) results. `Color` simply cannot be converted to `Rgb24` in a manner that preserves an alpha component without pre-multiplication and we currently have no pixel formats that do that multiplication. 

We should definitely treat those scenarios as separate concerns. 

@tocsoft  I do not want to encourage the non-generic API as our primary API. ImageSharp was envisioned to be different to (better than) System.Drawing in how pixel formats are handled. The whole point of the design was to allow users to make an explicit, educated, decision on what pixel format they used based upon their requirements. 

The baseline always was that users should be educated by the library so we don't get the copy/paste horror show that is 20 years of System.Drawing legacy.

I'm ok with a switch added to configuration but I remain unconvinced that it should default to true. `Rgba32` doesn't provide enough colorspace fidelity for many images so we would have to use `Rgba64` which means an explosion in memory usage.

**Comment by antonfirsov** [comment]
> We can fix that problem without breaking changes by adding a `FillColor` property to `ResizeOptions` and filling bounds within the padded sections of the image.

I agree that if we can improve this for Resize, we should.

> I do not want to encourage the non-generic API as our primary API

I think the major disagreements here are kinda philosophical, which makes them hard to overcome :) I *would* recommend the non-generic API as the primary one for everything that doesn't involve manual pixel processing or drawing.

Regardless of API & behavioral decisions, we need more/better docs on pixel formats and their implications on particular operations.

**Comment by br3aker** [comment]
> Rethinking the first point it would be possible using default interface implementations but I don't think that can be a V2 thing since any interface change would still be considered breaking.

But V2 is already released no? Anyway, next major version was target for this optimization in my head :)

> The whole point of the design was to allow users to make an explicit, educated, decision on what pixel format they used based upon their requirements.

This is a beutiful philosophy but I feel somebody has say it: sadly, there would always be people who just don't care or don't have resources/time/desire to study - they want results here and now while using existing library.

As I can see, all of us can agree on the 'parse with alpha channel if pixel type is not specified' flag set to whatever default value. Problem with setting it to 'parse without alpha' by default would produce 'invalid' results without any way of showing that in code/debug/release. IMO such problems should be shown by either code not compiling at all (which is tricky) then by debug exceptions and only then by actual docs. 

**I didn't really dig into the current image processing pipeline so below example might be completely irrelevant**

For code not compiling at all maybe a simple tag system like this can work:

```cs
public interface IPixel { }

internal interface INoAlphaTag { }

internal interface IHasAlphaTag { }

public struct Rgb24 : IPixel, INoAlphaTag { }

public struct Rgba32 : IPixel, IHasAlphaTag { }

public class Image<TPixel> where TPixel : IPixel { }

public static void RoundBorders<TPixel>(this Image<TPixel> image) where TPixel : IPixel, IHasAlphaTag { }

public static void RoundBorders<TPixel>(this Image<TPixel> image, TPixel fillCollor) where TPixel : IPixel, INoAlphaTag { }
```

P.S.
Agree with `FillColor` solution for Resize, good and simple.

**Comment by JimBobSquarePants** [comment]
Fixed via #2034
